core/server/proxy/src/main/java/alluxio/proxy/S3RestUtils.java

```diff
@@ -0,0 +1,283 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import javax.ws.rs.core.Response;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * Utilities for handling S3 REST calls.
+ */
+public final class S3RestUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(S3RestUtils.class);
+
+  /**
+   * Calls the given {@link S3RestUtils.RestCallable} and handles any exceptions thrown.
+   *
+   * @param <T> the return type of the callable
+   * @param resource the resource (bucket or object) to be operated on
+   * @param callable the callable to call
+   * @return the response object
+   */
+  public static <T> Response call(String resource, S3RestUtils.RestCallable<T> callable) {
+    try {
+      if (SecurityUtils.isSecurityEnabled() && AuthenticatedClientUser.get() == null) {
+        AuthenticatedClientUser.set(LoginUser.get().getName());
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to set AuthenticatedClientUser in REST service handler: {}", e.getMessage());
+      return createErrorResponse(new S3Exception(e, resource, ErrorCode.INTERNAL_ERROR));
+    }
+
+    try {
+      return createResponse(callable.call());
+    } catch (S3Exception e) {
+      LOG.warn("Unexpected error invoking REST endpoint: {}", e.getErrorCode().getDescription());
+      return createErrorResponse(e);
+    }
+  }
+
+  /**
+   * An interface representing a callable.
+   *
+   * @param <T> the return type of the callable
+   */
+  public interface RestCallable<T> {
+    /**
+     * The REST endpoint implementation.
+     *
+     * @return the return value from the callable
+     */
+    T call() throws S3Exception;
+  }
+
+  /**
+   * Creates a response using the given object.
+   *
+   * @param object the object to respond with
+   * @param resource the resource name (bucket or object key)
+   * @return the response
+   */
+  private static Response createResponse(Object object) {
+    if (object instanceof Void) {
+      return Response.ok().build();
+    }
+    return Response.ok(object).build();
+  }
+
+  /**
+   * Creates an error response using the given exception.
+   *
+   * @param e the exception to be converted into {@link Error} and encoded into XML
+   * @return the response
+   */
+  private static Response createErrorResponse(S3Exception e) {
+    Error errorResponse = new Error(e.getResource(), e.getErrorCode());
+    return Response.status(e.getErrorCode().getStatus()).entity(errorResponse).build();
+  }
+
+  /**
+   * Error code names used in {@link ErrorCode}.
+   */
+  public static class ErrorCodeName {
+    public static String BUCKET_ALREADY_EXISTS = "BucketAlreadyExists";
+    public static String INTERNAL_ERROR = "InternalError";
+    public static String INVALID_BUCKET_NAME = "InvalidBucketName";
+  }
+
+  /**
+   * Error codes defined in http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html,
```

Comment:
any plan for how to make this always in sync with `latest` AWS S3 API? or shall we mark this comment to a specific API version?